Enable edition 2018

tcharding opened the pull request:
Enable edition 2018. While we are at it do a whole bunch of cleanups. This PR is basically everything that I've been doing stack wide but in a single PR. The overall flow of the PR is:

- Remove a single deprecation warning
- Changed the example in the readme
- Enabled edition 2018
- Cleared _all_ clippy warnings
- Added clippy to CI
- Add pre-commit githook including clippy
- Implement `std::error::Error` crate wide

tcharding commented:
Legend, thanks @dpc!

tcharding commented:
Can you kick your CI bot for me please @apoelstra?

apoelstra commented:
Nice! I notice though that the library doesn't actually compile now with some feature combinations (specifically, `simple_http` but not `std`). Can we make `simple_http` depend on `std`?

tcharding commented:
My apologies, the new `std` feature was half baked. This crate uses `std` all over the  place. I removed the new `std` feature and just implement `std::error::Error` unconditionally now.

Also added a patch to remove the `base64-compat` dependency and use plain `base64` same as we did in `rust-bitcoin`.

apoelstra commented:
The commit message for 94dce94 still describes adding a std feature.

apoelstra reviewed (approved):
ACK 982e37ac912b6ae304cc413aaecaf05a20ac096e

tcharding commented:
> The commit message for https://github.com/apoelstra/rust-jsonrpc/commit/94dce94eee2e80a1c19dc151709053271b62a431 still describes adding a std feature.

Drats, my bad. Will fix and re-spin.

tcharding commented:
Force push is fix to the commit log of commit `162d4a2 Implement std::error::Error for all errors` to remove mention of `std` feature. No other changes.
